experimental/auth-logic/BUILD

```diff
@@ -0,0 +1,53 @@
+#
+# Copyright 2022 The Project Oak Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test", "go_binary")
+
+package(default_visibility = ["//:__subpackages__"])
+
+go_library(
+  name = "auth_logic_interface",
+  importpath = "github.com/project-oak/transparent-release/experimental/auth-logic",
+  srcs = [
+    "auth_logic_interface.go",
+  ],
+  data = [
+    "@auth-logic-compiler//file",
+    "simple.auth_logic"
+  ],
+)
+
+go_test(
+    name = "simple_auth_logic_test",
+    size = "small",
+    srcs = ["simple_auth_logic_test.go"],
+    embed = [":auth_logic_interface"],
+    data = [
+      "demo_working.csv",
+      "demo_disappointing.csv",
+    ],
+)
+
+genrule(
+  name = "simple_auth_logic_compiler",
+  srcs = ["simple.auth_logic"],
+  tools = ["@auth-logic-compiler//file"],
+  cmd = """
+    export INPUT_DIR=$$(dirname $(location simple.auth_logic))
+    ./$(location @auth-logic-compiler//file) simple.auth_logic $$INPUT_DIR $(RULEDIR)
+  """,
+  outs = ["simple.auth_logic.dl", "demo_working.csv", "demo_disappointing.csv"]
+)
```

Comment:
I'm not sure this genrule actually buys us anything. In fact it seems it makes it harder / impossible to evaluate policies at runtime (e.g. in a Go test), which I think is something that will be useful. For instance the current logic is split between this genrule and some logic in the Go file that parses the output of this genrule at runtime. I would expect the top-level binary / function to accept the actual policy as input (i.e. a literal string, which may be read from a file by the caller), and return the map[string]bool (which it currently does already, so that's fine). That function will invoke the auth logic compiler and souffle behind the scenes, but callers should not need to know that, and should not need to use genrules themselves.

Reply:
A bazel rule for running the authorization logic compiler is needed in order to run it in a `go_test` because just finding the binary relative to where the bazel sandbox is executing from within go (which is what I was doing before) is too brittle to even work in both the `go_binary` and `go_test` rules because the paths change.

Since authorization logic is a programming language, having a bazel rule to run it seems like the most natural way to integrate it with Bazel.

It seems like your complaint is just that it runs some `go` code after it runs the auth logic compiler in order to make the results from auth logic interpretable to a `go` test. I don't know why you see this as a problem. I think any time you have a program split between two programming languages you will need to involve the compiler for both languages.

I think for the release process, the flow is that:
 - the top level comprises a bunch of go functions that parse various files and emit authorization logic
 - the authorization logic compiler runs on the generated authorization logic code to generate souffle
 - souffle is used to evaluate a query that decides if the binary is accepted or not

Since the last step is just a single decision, I think it doesn't even matter if this final decision is interpreted from `go` or not. So actually passing results from authorization logic back to go is not even strictly needed (but if that is wanted, this gives a rough start to that). The point of this PR is really just to make sure the authorization logic compiler can be run from within this repo.

Reply:
I actually was trying to just invoke the binary from within bazel code and run it in a `go_test` rule but that already results in the auth logic binary being placed in a different path from where it is placed in a `go_binary` environment.

Reply:
Maybe the reason you don't like this is that the specific input program to the compiler (`simple.auth_logic`) is hard-coded rather than a variable that gets passed to the compiler (and same for the output queries)? I intend to make a more reusable build rule for the auth logic compiler, but I am new to both bazel and go.

Reply:
I am also still trying to get this to work by using the auth logic compiler as a binary, so I probably won't do this refactoring until it works with a `go_test` in the first place.

Reply:
Thanks, yes I think your points address my doubts, but we need to have a more solid way of specifying what the output should be from the combined auth logic + souffle stack, I don't think we can expect callers to parse the CSV files, so perhaps the genrule can evolve to actually include the parsing logic that is currently in your Go file (compiled to another binary?), and we would come up with some better format for the actual output; e.g. a JSON file with the list of predicates and their values, similar to the map[string]bool? or something else that we can more easily use to drive additional tools downstream, e.g. signing a particular binary if the output of the auth logic compiler for a particular statement is true (which I imagine is one of our primary use cases for the release process).

Reply:
In [Section 3.5, the Eng plan](https://docs.google.com/document/d/1eKHhiPonEbBqkba9UIlpUXFVVBgLiRI_QDT3CJwFLO4/edit#bookmark=id.rarevgu5bpm3) for this milestone had evaluating a single query as part of the CLI. (I realize internal docs are not the most open-source friendly :-/) That isn't done yet though. Does that way still work?

Another thing I was thinking if CSVs are kept around at all is to just have a single --output_query_csv <filename> that results in just one CSV with the results for all of the named queries. (Seems like at that point a single CSV can be converted into any other format the consumer of this thing desires).

Reply:
I think the current logic is fine, but I want to make sure that eventually (i.e. in future PRs) the CSV parsing logic should also become part of the bazel genrule (or some more typesafe bazel rule), and the output should be something more "standard", like a JSON file or similar, with a well defined schema, that callers can consume without having to deal with the impl details of datalog.

Reply:
https://github.com/google-research/raksha/issues/455